### PR TITLE
Add Merge PullRequest API

### PR DESCRIPTION
**Pull request by oshiro3**
I create a API according to the following specifications.
https://developer.github.com/v3/pulls/#merge-a-pull-request

`SHA` and `COMMIT_TITLE` parameters are not supported.

But I have one question here.
If `request.body` is passed empty, it was not be handled and an error occured with 500: InternalServerError. It is returned at that point.
I want advice on this issue.

### Before submitting a pull-request to GitBucket I have first:

- [x] read the [contribution guidelines](https://github.com/gitbucket/gitbucket/blob/master/.github/CONTRIBUTING.md)
- [ ] rebased my branch over master
- [x] verified that project is compiling
- [x] verified that tests are passing
- [x] squashed my commits as appropriate *(keep several commits if it is relevant to understand the PR)*
- [ ] [marked as closed using commit message](https://help.github.com/articles/closing-issues-via-commit-messages/) all issue ID that this PR should correct

### CODE REVIEW

**Comment by takezoe** [comment]
@oshiro3 Maybe that is a general behavior in APIs which use `extractFromJsonBody`, though, needs to be improved. Do you have a stacktrace?

**Comment by oshiro3** [comment]
@takezoe 
Thank you for watching PR.
I figured out whether to modify the `extractFromJsonBody` itself or handle the empty body in the controller.
However, in general, I also wondered if the body using the POST or PUT method would rarely be empty, so I could not determine whether `extractFromJsonBody` itself should handle an empty body.

I send the following request.
`curl -u Oshiro:pass localhost:8080/api/v3/repos/Oshiro/Test/pulls/1/merge -X PUT`
This is gitbucket log.
```2020-07-13 19:48:40.048:INFO:oejs.Server:main: Started @6485ms
19:49:55.546 [qtp1144748369-13] INFO  o.c.core.Cache2kCoreProviderImpl - cache2k starting. version=1.2.4.Final
19:49:55.611 [qtp1144748369-13] WARN  o.c.e.a.WarnAboutCache2kAllUsage - Usage of the artifact cache2k-all is deprecated, please update your dependencies. See: https://cache2k.org/docs/latest/user-guide.html#getting-started
19:49:56.248 [qtp1144748369-13] ERROR g.core.controller.ApiController - Catch unhandled error in request: Request(PUT //localhost:8080/api/v3/repos/Oshiro/Test/pulls/1/merge)@316a8a1
com.fasterxml.jackson.databind.exc.MismatchedInputException: No content to map due to end-of-input
 at [Source: (String)""; line: 1, column: 0]
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:59)
	at com.fasterxml.jackson.databind.DeserializationContext.reportInputMismatch(DeserializationContext.java:1356)
	at com.fasterxml.jackson.databind.ObjectReader._initForReading(ObjectReader.java:358)
	at com.fasterxml.jackson.databind.ObjectReader._bindAndClose(ObjectReader.java:1596)
	at com.fasterxml.jackson.databind.ObjectReader.readValue(ObjectReader.java:1219)
	at org.json4s.jackson.JsonMethods.parse(JsonMethods.scala:25)
	at org.json4s.jackson.JsonMethods.parse$(JsonMethods.scala:19)
	at gitbucket.core.controller.ControllerBase.parse(ControllerBase.scala:35)
	at gitbucket.core.controller.ControllerBase.extractFromJsonBody(ControllerBase.scala:232)
	at gitbucket.core.controller.api.ApiPullRequestControllerBase.$anonfun$$init$$34(ApiPullRequestControllerBase.scala:226)
	at gitbucket.core.util.ReferrerAuthenticator.$anonfun$authenticate$4(Authenticator.scala:95)
	at scala.Option.map(Option.scala:242)
	at gitbucket.core.util.ReferrerAuthenticator.authenticate(Authenticator.scala:93)
	at gitbucket.core.util.ReferrerAuthenticator.referrersOnly(Authenticator.scala:87)
	at gitbucket.core.util.ReferrerAuthenticator.referrersOnly$(Authenticator.scala:87)
	at gitbucket.core.controller.ApiController.referrersOnly(ApiController.scala:10)
	at gitbucket.core.controller.api.ApiPullRequestControllerBase.$anonfun$$init$$33(ApiPullRequestControllerBase.scala:222)
	at org.scalatra.ScalatraBase.liftAction(ScalatraBase.scala:282)
	at org.scalatra.ScalatraBase.$anonfun$invoke$1(ScalatraBase.scala:276)
	at org.scalatra.ApiFormats.withRouteMultiParams(ApiFormats.scala:185)
	at org.scalatra.ApiFormats.withRouteMultiParams$(ApiFormats.scala:171)
	at gitbucket.core.controller.ControllerBase.withRouteMultiParams(ControllerBase.scala:35)
	at org.scalatra.ScalatraBase.invoke(ScalatraBase.scala:275)
	at org.scalatra.ScalatraBase.invoke$(ScalatraBase.scala:274)
	at gitbucket.core.controller.ControllerBase.invoke(ControllerBase.scala:35)
	at org.scalatra.ScalatraBase.$anonfun$runRoutes$3(ScalatraBase.scala:250)
	at scala.Option.flatMap(Option.scala:283)
	at org.scalatra.ScalatraBase.$anonfun$runRoutes$1(ScalatraBase.scala:248)
	at scala.collection.immutable.Stream.flatMap(Stream.scala:206)
	at org.scalatra.ScalatraBase.runRoutes(ScalatraBase.scala:247)
	at org.scalatra.ScalatraBase.runRoutes$(ScalatraBase.scala:245)
	at gitbucket.core.controller.ControllerBase.runRoutes(ControllerBase.scala:35)
	at org.scalatra.ScalatraBase.runActions$1(ScalatraBase.scala:169)
	at org.scalatra.ScalatraBase.$anonfun$executeRoutes$6(ScalatraBase.scala:181)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at org.scalatra.ScalatraBase.cradleHalt(ScalatraBase.scala:201)
	at org.scalatra.ScalatraBase.executeRoutes(ScalatraBase.scala:181)
	at org.scalatra.ScalatraBase.executeRoutes$(ScalatraBase.scala:150)
	at gitbucket.core.controller.ControllerBase.executeRoutes(ControllerBase.scala:35)
	at org.scalatra.ScalatraBase.$anonfun$handle$1(ScalatraBase.scala:123)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.DynamicScope.withResponse(DynamicScope.scala:75)
	at org.scalatra.DynamicScope.withResponse$(DynamicScope.scala:73)
	at gitbucket.core.controller.ControllerBase.withResponse(ControllerBase.scala:35)
	at org.scalatra.DynamicScope.$anonfun$withRequestResponse$1(DynamicScope.scala:55)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.DynamicScope.withRequest(DynamicScope.scala:66)
	at org.scalatra.DynamicScope.withRequest$(DynamicScope.scala:64)
	at gitbucket.core.controller.ControllerBase.withRequest(ControllerBase.scala:35)
	at org.scalatra.DynamicScope.withRequestResponse(DynamicScope.scala:54)
	at org.scalatra.DynamicScope.withRequestResponse$(DynamicScope.scala:52)
	at gitbucket.core.controller.ControllerBase.withRequestResponse(ControllerBase.scala:35)
	at org.scalatra.ScalatraBase.handle(ScalatraBase.scala:123)
	at org.scalatra.ScalatraBase.handle$(ScalatraBase.scala:119)
	at gitbucket.core.controller.ControllerBase.org$scalatra$servlet$ServletBase$$super$handle(ControllerBase.scala:35)
	at org.scalatra.servlet.ServletBase.handle(ServletBase.scala:42)
	at org.scalatra.servlet.ServletBase.handle$(ServletBase.scala:35)
	at gitbucket.core.controller.ControllerBase.org$scalatra$FlashMapSupport$$super$handle(ControllerBase.scala:35)
	at org.scalatra.FlashMapSupport.$anonfun$handle$1(FlashMap.scala:197)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.DynamicScope.withRequest(DynamicScope.scala:66)
	at org.scalatra.DynamicScope.withRequest$(DynamicScope.scala:64)
	at gitbucket.core.controller.ControllerBase.withRequest(ControllerBase.scala:35)
	at org.scalatra.FlashMapSupport.handle(FlashMap.scala:172)
	at org.scalatra.FlashMapSupport.handle$(FlashMap.scala:171)
	at gitbucket.core.controller.ControllerBase.handle(ControllerBase.scala:35)
	at org.scalatra.ScalatraFilter.$anonfun$doFilter$1(ScalatraFilter.scala:41)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.ScalatraFilter.doFilter(ScalatraFilter.scala:41)
	at org.scalatra.ScalatraFilter.doFilter$(ScalatraFilter.scala:36)
	at gitbucket.core.controller.ControllerBase.doFilter(ControllerBase.scala:73)
	at gitbucket.core.servlet.CompositeScalatraFilter.$anonfun$process$2(CompositeScalatraFilter.scala:67)
	at gitbucket.core.servlet.CompositeScalatraFilter.$anonfun$process$2$adapted(CompositeScalatraFilter.scala:64)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:553)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:551)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:921)
	at gitbucket.core.servlet.CompositeScalatraFilter.process(CompositeScalatraFilter.scala:64)
	at gitbucket.core.servlet.ControllerFilter.doFilter(CompositeScalatraFilter.scala:25)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1591)
	at gitbucket.core.servlet.ControllerFilter.doFilter(CompositeScalatraFilter.scala:31)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1591)
	at org.scalatra.ScalatraFilter.$anonfun$doNotFound$1(ScalatraFilter.scala:79)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at org.scalatra.ScalatraBase.$anonfun$executeRoutes$5(ScalatraBase.scala:172)
	at scala.Option.getOrElse(Option.scala:201)
	at org.scalatra.ScalatraBase.$anonfun$executeRoutes$3(ScalatraBase.scala:172)
	at scala.Option.getOrElse(Option.scala:201)
	at org.scalatra.ScalatraBase.runActions$1(ScalatraBase.scala:172)
	at org.scalatra.ScalatraBase.$anonfun$executeRoutes$6(ScalatraBase.scala:181)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at org.scalatra.ScalatraBase.cradleHalt(ScalatraBase.scala:201)
	at org.scalatra.ScalatraBase.executeRoutes(ScalatraBase.scala:181)
	at org.scalatra.ScalatraBase.executeRoutes$(ScalatraBase.scala:150)
	at gitbucket.core.controller.ControllerBase.executeRoutes(ControllerBase.scala:35)
	at org.scalatra.ScalatraBase.$anonfun$handle$1(ScalatraBase.scala:123)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.DynamicScope.withResponse(DynamicScope.scala:75)
	at org.scalatra.DynamicScope.withResponse$(DynamicScope.scala:73)
	at gitbucket.core.controller.ControllerBase.withResponse(ControllerBase.scala:35)
	at org.scalatra.DynamicScope.$anonfun$withRequestResponse$1(DynamicScope.scala:55)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.DynamicScope.withRequest(DynamicScope.scala:66)
	at org.scalatra.DynamicScope.withRequest$(DynamicScope.scala:64)
	at gitbucket.core.controller.ControllerBase.withRequest(ControllerBase.scala:35)
	at org.scalatra.DynamicScope.withRequestResponse(DynamicScope.scala:54)
	at org.scalatra.DynamicScope.withRequestResponse$(DynamicScope.scala:52)
	at gitbucket.core.controller.ControllerBase.withRequestResponse(ControllerBase.scala:35)
	at org.scalatra.ScalatraBase.handle(ScalatraBase.scala:123)
	at org.scalatra.ScalatraBase.handle$(ScalatraBase.scala:119)
	at gitbucket.core.controller.ControllerBase.org$scalatra$servlet$ServletBase$$super$handle(ControllerBase.scala:35)
	at org.scalatra.servlet.ServletBase.handle(ServletBase.scala:42)
	at org.scalatra.servlet.ServletBase.handle$(ServletBase.scala:35)
	at gitbucket.core.controller.ControllerBase.org$scalatra$FlashMapSupport$$super$handle(ControllerBase.scala:35)
	at org.scalatra.FlashMapSupport.$anonfun$handle$1(FlashMap.scala:197)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.DynamicScope.withRequest(DynamicScope.scala:66)
	at org.scalatra.DynamicScope.withRequest$(DynamicScope.scala:64)
	at gitbucket.core.controller.ControllerBase.withRequest(ControllerBase.scala:35)
	at org.scalatra.FlashMapSupport.handle(FlashMap.scala:172)
	at org.scalatra.FlashMapSupport.handle$(FlashMap.scala:171)
	at gitbucket.core.controller.ControllerBase.handle(ControllerBase.scala:35)
	at org.scalatra.ScalatraFilter.$anonfun$doFilter$1(ScalatraFilter.scala:41)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:59)
	at org.scalatra.ScalatraFilter.doFilter(ScalatraFilter.scala:41)
	at org.scalatra.ScalatraFilter.doFilter$(ScalatraFilter.scala:36)
	at gitbucket.core.controller.ControllerBase.doFilter(ControllerBase.scala:73)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1591)
	at gitbucket.core.servlet.ApiAuthenticationFilter.doFilter(ApiAuthenticationFilter.scala:43)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1591)
	at gitbucket.core.servlet.TransactionFilter.$anonfun$doFilter$1(TransactionFilter.scala:39)
	at gitbucket.core.servlet.TransactionFilter.$anonfun$doFilter$1$adapted(TransactionFilter.scala:30)
	at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.$anonfun$withTransaction$2(BlockingProfile.scala:207)
	at slick.JdbcProfileBlockingSession$BlockingSession.withTransaction(TransactionalJdbcBackend.scala:26)
	at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.$anonfun$withTransaction$1(BlockingProfile.scala:207)
	at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.withSession(BlockingProfile.scala:200)
	at com.github.takezoe.slick.blocking.BlockingJdbcProfile$BlockingAPI$BlockingDatabase.withTransaction(BlockingProfile.scala:207)
	at gitbucket.core.servlet.TransactionFilter.doFilter(TransactionFilter.scala:30)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1583)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:542)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:536)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:235)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1581)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1307)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:482)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1549)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1204)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:221)
	at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:146)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:494)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:374)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:268)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:103)
	at org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:117)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:782)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:918)
	at java.lang.Thread.run(Thread.java:748)
``` #

**Comment by takezoe** [comment]
Thanks! I will take a closer look when I can.

**Comment by oshiro3** [comment]
＠Takezoe, how about this problem? If we need to fix the `extractFromJsonBody` function, it is not a problem that can be handled by this PR, so I would like to fix other problems and proceed. 
Thank you.

**Comment by takezoe** [comment]
Oops. Sorry, I've not looked at this yet. However, I agree that fixing `extractFromJsonBody` is another issue.

**Comment by oshiro3** [comment]
@takezoe 
OK!! I updated  API wiki page.

File: src/main/scala/gitbucket/core/api/MergeAPullRequest.scala [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+)
+
+case class SuccessToMergePrResponse(
+  sha: String,
+  merged: Boolean,
+  message: String
+)
+
+case class FailToMergePrResponse(
+  documentation_ur: String,
```
Comment (SIkebe) [inline src/main/scala/gitbucket/core/api/MergeAPullRequest.scala:21]: `documentation_url` ?

File: src/main/scala/gitbucket/core/api/MergeAPullRequest.scala [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package gitbucket.core.api
+
+/**
+ * https://developer.github.com/v3/pulls/#merge-a-pull-request
```
Comment (SIkebe) [inline src/main/scala/gitbucket/core/api/MergeAPullRequest.scala:4]: Hmm, this site is now seem to be outdated.
@takezoe Is it better to use https://docs.github.com/en/rest/reference/pulls#merge-a-pull-request-merge-button intead?

Comment (SIkebe) [reply src/main/scala/gitbucket/core/api/MergeAPullRequest.scala:4]: Sorry, this is the right URL 😅 
https://docs.github.com/en/rest/reference/pulls#merge-a-pull-request

File: src/main/scala/gitbucket/core/api/MergeAPullRequest.scala (hunk 1 above)
Comment (takezoe) [reply src/main/scala/gitbucket/core/api/MergeAPullRequest.scala:21]: @oshiro3 Could you take a look at this?

Comment (oshiro3) [reply src/main/scala/gitbucket/core/api/MergeAPullRequest.scala:21]: yes. I fix this typo.

@SIkebe 
Thanks for finding the typo.

File: src/main/scala/gitbucket/core/api/MergeAPullRequest.scala (hunk 2 above)
Comment (oshiro3) [reply src/main/scala/gitbucket/core/api/MergeAPullRequest.scala:4]: OK. I replace with a new URL.

Comment (oshiro3) [reply src/main/scala/gitbucket/core/api/MergeAPullRequest.scala:4]: @SIkebe Thank you for confirming!!

**Comment by SIkebe** [comment]
Need test here?
https://github.com/gitbucket/gitbucket/blob/master/src/test/scala/gitbucket/core/api/JsonFormatSpec.scala

**Comment by oshiro3** [comment]
> Need test here?
> https://github.com/gitbucket/gitbucket/blob/master/src/test/scala/gitbucket/core/api/JsonFormatSpec.scala

I confirmed the test.
I thought the MergeAPullRequest class is different from the API model listed here and I may not need to write a test.
I think that MergeAPullRequest class is the same kind as CreateXxx.
Also, I think I need to implement object just for testing, but what about?

File: src/main/scala/gitbucket/core/controller/api/ApiPullRequestControllerBase.scala [hunk 3]
```diff
@@ -217,8 +217,72 @@ trait ApiPullRequestControllerBase extends ControllerBase {
...
+                    message = "Pull Request successfully merged"
+                  )
+                )
+              )
+            case Left(message) =>
+              MethodNotAllowed(
+                JsonFormat(
+                  FailToMergePrResponse(
+                    message = "Pull Request is not mergeable",
+                    documentation_url = "https://developer.github.com/v3/pulls/#merge-a-pull-request-merge-button",
```
Comment (SIkebe) [inline src/main/scala/gitbucket/core/controller/api/ApiPullRequestControllerBase.scala:277]:
```suggestion
                    documentation_url = "https://docs.github.com/en/rest/reference/pulls#merge-a-pull-request",
```

**Comment by SIkebe** [comment]
>I thought the MergeAPullRequest class is different from API models listed here and I may not need to add a test.

@oshiro3 Yeah, you're right. Thanks!

File: src/main/scala/gitbucket/core/controller/api/ApiPullRequestControllerBase.scala (hunk 3 above)
Comment (oshiro3) [reply src/main/scala/gitbucket/core/controller/api/ApiPullRequestControllerBase.scala]: replace the old doc url.
thank you!

**Comment by oshiro3** [comment]
@SIkebe 
I am very pleased with your review.
Thank you!!

**Comment by takezoe** [comment]
Merged. Thanks for both you guys!

**Pull request merged**